controller/src/beerocks/bml/internal/bml_internal.cpp

```diff
@@ -829,6 +832,48 @@ int bml_internal::process_cmdu_header(cmdu_vs_action_header_t beerocks_header,
             *m_pvaps_list_size = i;
 
         } break;
+        case beerocks_message::ACTION_BML_WFA_CA_CONTROLLER_RESPONSE: {
+            auto response =
+                cmdu_rx.addClass<beerocks_message::cACTION_BML_WFA_CA_CONTROLLER_RESPONSE>();
+            if (!response) {
+                LOG(ERROR) << "addClass cACTION_BML_WFA_CA_CONTROLLER_RESPONSE failed";
+                return BML_RET_OP_FAILED;
+            }
+
+            auto reply_tuple = response->reply(0);
+            if (!std::get<0>(reply_tuple)) {
+                LOG(ERROR) << "getting reply entry has failed!";
+                return false;
+            }
+
+            char *reply_ptr = (char *)(&std::get<1>(reply_tuple));
+
+            // first reply
+            if (!m_wfa_ca_controller_pending_second_reply) {
+                m_wfa_ca_controller_first_reply.assign(reply_ptr);
+                m_wfa_ca_controller_pending_second_reply = response->pending_second_response();
+                // Signal any waiting threads
+                if (!wake_up(beerocks_message::ACTION_BML_WFA_CA_CONTROLLER_REQUEST, true)) {
+                    LOG(WARNING) << "Received ACTION_BML_WFA_CA_CONTROLLER_REQUEST response, "
+                                    "but no one is waiting...";
+                }
+
+                // Must use sleep here to prevent a race between the first reply to the second reply
+                // Without sleep here, the second reply may be printed first instead of the first
+                // reply, if both replies were sent at the same time.
```

Comment:
Sleeping is never a solution for a race condition, it is wiping it under the carpet so it can show up at the worst possible time.

By always printing results in the callback, avoid this issue entirely because you never need to synchronise the command with the response.